### ISSUE TITLE
feat: pause peer timeouts during longer block execution

### ISSUE THREAD

**Issue by czarcas7ic** [cometbft/cometbft#2533]
### Summary

Currently, when processing large blocks (epoch for Osmosis), we time out peers despite the fact that we don't expect anything from them during this time. 

### Problem Definition

Especially with the P2P improvements, we now rank peers based on their speed of sending blocks to us. When we get to epoch and start disconnecting from perfectly good peers, it takes some time to get back to an ideal state.

### Proposal

It would be nice if peer timeout logic could detect if we are in the process of processing a block still, and not time out peers during this time.

**Comment by cason** [comment]
Not sure if I understood which protocol you are referring to: consensus, PEX, other?

**Comment by czarcas7ic** [comment]
My apologies for not being very specific here. These are the logs we get while our network processes a ~2min long epoch.

![Screenshot 2024-03-09 at 12 37 07 PM](https://github.com/cometbft/cometbft/assets/40078083/bc0c5fc1-001e-4182-8ce8-1939278d358c)

The desire here is that if we are caught up to the head of the chain and then are still in the middle of processing a larger block, we don't time out peers because we don't expect any peers to have anything to respond with during this time. Please let me know if this clarifies!

**Comment by cason** [comment]
A possibly reason for that, and this should not happen in released versions, is that this blocking processing step is happening as part of the `Receive` method of a reactor. This method should not block. If it does, the main routine of the multiplex transport also blocks, which is not great.

**Comment by czarcas7ic** [comment]
Hey @cason sorry for the delayed response here. We can replicate this behavior on every epoch, so am happy to provide you any further information that might help with debugging this.

**Comment by czarcas7ic** [comment]
Epoch lasts approx 3minutes, so 45s is much lower than this, but not sure if your statement is that no matter how long the block is, the pong should still occur.

**Comment by cason** [comment]
> but not sure if your statement is that no matter how long the block is, the pong should still occur.

It should, as it is an independent routine. Namely, the connection with a peer should not become stale because the node, or that peer, are busy doing something else. The ping-pong messages are intended for that, they should work like an application level "keepalive" message.

**Comment by cason** [comment]
For some reason, I cannot open the image so to read the logs better.

Moreover, we need to know if there are code changes in the p2p package, or the original version is in use.

**Comment by czarcas7ic** [comment]
The logs just essentially say stopping peer for error, pong timeout for a bunch of different peer IDs. We are using v0.37.4, with some cherry picked commits from v0.37.x, but nothing super custom.

https://github.com/osmosis-labs/cometbft/tree/osmo/v0.37.4

**Comment by cason** [comment]
We need to understand which reactor is blocking in the `Receive()` call. My guess is the mempool reactor. If transactions are still received but `CheckTx` is not being performed and returned by the ABCI application, we could have the described scenario. Is the core of the processing step implemented by the `Commit()` ABCI call?

**Comment by czarcas7ic** [comment]
If I recall correctly, these messages come before the commit step is even called

**Comment by cason** [comment]
So, we cannot and should not change p2p timeouts when we are running a processing task. This because the main routine of a p2p connection should not block, even if we are running a CPU heavy task.

The p2p connection main routine may block (this is to be investigate) when the `Receive()` call at some reactor blocks. The implementation of these call should not block. We need to investigate which reactor may be blocking, for that long period, during this call. And fix the behavior of that reactor.

In other words, this is a more complex problem that we should, in my opinion, address in a different issue.

**Comment by cason** [comment]
Created #2685 to handle this problem in a more generic way.

**Comment by czarcas7ic** [comment]
Great thank you! Feel free to close this issue then if you feel the new issue is all that is needed to track!

**Comment by cason** [comment]
Issue https://github.com/cometbft/cometbft/issues/2685 should tackle the problem reported on this issue. Closing.
